test_util/test_upgrade_vpc.py

```diff
@@ -159,125 +132,277 @@ def wait_for_dns(hostname):
-        #raise Exception(message)  # noqa
+
+class VpcClusterUpdateTestDcosApiSessionFactory:
+    def apply(self, dcos_url: str, masters: List[str], public_masters: List[str], slaves: List[str],
+              public_slaves: List[str], default_os_user: str) -> DcosApiSession:
+        pass
+
+
+class VpcClusterUpgradeTest:
+    log = logging.getLogger(__name__)
+
+    def __init__(self,
+                 num_masters: int, num_agents: int, num_public_agents: int,
+                 stable_installer_url: str, installer_url: str,
+                 aws_region: str, aws_access_key_id: str, aws_secret_access_key: str,
+                 ssh_key: str, default_os_user: str,
+                 config_yaml_override_install: str, config_yaml_override_upgrade: str,
+                 dcos_api_session_factory_install: VpcClusterUpdateTestDcosApiSessionFactory,
+                 dcos_api_session_factory_upgrade: VpcClusterUpdateTestDcosApiSessionFactory):
+        self.dcos_api_session_factory_install = dcos_api_session_factory_install
+        self.dcos_api_session_factory_upgrade = dcos_api_session_factory_upgrade
+        self.num_masters = num_masters
+        self.num_agents = num_agents
+        self.num_public_agents = num_public_agents
+        self.stable_installer_url = stable_installer_url
+        self.installer_url = installer_url
+        self.aws_region = aws_region
+        self.aws_access_key_id = aws_access_key_id
+        self.aws_secret_access_key = aws_secret_access_key
+        self.ssh_key = ssh_key
+        self.default_os_user = default_os_user
+        self.config_yaml_override_install = config_yaml_override_install
+        self.config_yaml_override_upgrade = config_yaml_override_upgrade
+
+        self.teamcity_msg = TeamcityServiceMessages()
+
+        # the two following properties are set when running setup_cluster_workload, here we default them to empty
+        # values.
+        self.test_app_ids = []
+        self.tasks_start = []
+
+    @staticmethod
+    def app_task_ids(dcos_api, app_id):
+        """Return a list of Mesos task IDs for app_id's running tasks."""
+        assert app_id.startswith('/')
+        response = dcos_api.marathon.get('/v2/apps' + app_id + '/tasks')
+        response.raise_for_status()
+        tasks = response.json()['tasks']
+        return [task['id'] for task in tasks]
+
+    def parse_dns_log(self, dns_log_content):
+        """Return a list of (timestamp, status) tuples from dns_log_content."""
+        dns_log = [line.strip().split(' ') for line in dns_log_content.strip().split('\n')]
+        if any(len(entry) != 2 or entry[1] not in ['SUCCESS', 'FAILURE'] for entry in dns_log):
+            message = 'Malformed DNS log.'
+            self.log.debug(message + ' DNS log content:\n' + dns_log_content)
+            raise Exception(message)
+        return dns_log
+
+    def log_test(self, test_name: str, call: Callable[[], None]) -> None:
```

Comment:
I'll look into the decorator option.

I don't really want to centralize this too much, since I would prefer that we get to the point that instead of doing this we are able to write pytests that are aware of the lifecycle of a cluster being upgraded.